kornia/feature/affine_shape.py

```diff
@@ -120,3 +126,93 @@ def forward(self, laf: torch.Tensor, img: torch.Tensor) -> torch.Tensor:  # type
+    then runs AffNet on patches to estimate LAFs shape.
+    This is based on the original code from paper "Repeatability Is Not Enough:
+    Learning Discriminative Affine Regions via Discriminability"".
+    See :cite:`AffNet2018` for more details.
+    Then original LAF shape is replaced with estimated one. The original LAF orientation is not preserved,
+    so it is recommended to first run LAFAffineShapeEstimator and then LAFOrienter.
+
+    Args:
+        pretrained: (bool) Download and set pretrained weights to the model. Default: false.
+    """
+    def __init__(self, pretrained: bool = False):
+        super(LAFAffNetShapeEstimator, self).__init__()
+        self.features = nn.Sequential(
+            nn.Conv2d(1, 16, kernel_size=3, padding=1, bias=False),
+            nn.BatchNorm2d(16, affine=False),
+            nn.ReLU(),
+            nn.Conv2d(16, 16, kernel_size=3, stride=1, padding=1, bias=False),
+            nn.BatchNorm2d(16, affine=False),
+            nn.ReLU(),
+            nn.Conv2d(16, 32, kernel_size=3, stride=2, padding=1, bias=False),
+            nn.BatchNorm2d(32, affine=False),
+            nn.ReLU(),
+            nn.Conv2d(32, 32, kernel_size=3, stride=1, padding=1, bias=False),
+            nn.BatchNorm2d(32, affine=False),
+            nn.ReLU(),
+            nn.Conv2d(32, 64, kernel_size=3, stride=2, padding=1, bias=False),
+            nn.BatchNorm2d(64, affine=False),
+            nn.ReLU(),
+            nn.Conv2d(64, 64, kernel_size=3, stride=1, padding=1, bias=False),
+            nn.BatchNorm2d(64, affine=False),
+            nn.ReLU(),
+            nn.Dropout(0.25),
+            nn.Conv2d(64, 3, kernel_size=8, stride=1, padding=0, bias=True),
+            nn.Tanh(),
+            nn.AdaptiveAvgPool2d(1)
+        )
+        self.patch_size = 32
+        # use torch.hub to load pretrained model
+        if pretrained:
+            pretrained_dict = torch.hub.load_state_dict_from_url(
+                urls['affnet'], map_location=lambda storage, loc: storage
+            )
+            self.load_state_dict(pretrained_dict['state_dict'], strict=False)
+        return
+
+    @staticmethod
+    def _normalize_input(x: torch.Tensor, eps: float = 1e-6) -> torch.Tensor:
+        "Utility function that normalizes the input by batch."""
+        sp, mp = torch.std_mean(x, dim=(-3, -2, -1), keepdim=True)
+        # WARNING: we need to .detach() input, otherwise the gradients produced by
+        # the patches extractor with F.grid_sample are very noisy, making the detector
+        # training totally unstable.
+        return (x - mp.detach()) / (sp.detach() + eps)
+
+    def forward(self, laf: torch.Tensor, img: torch.Tensor) -> torch.Tensor:  # type: ignore
+        """
+        Args:
+            laf: (torch.Tensor) shape [BxNx2x3]
+            img: (torch.Tensor) shape [Bx1xHxW]
+
+        Returns:
+            laf_out: (torch.Tensor) shape [BxNx2x3]"""
```

Comment:
this won't render well I believe. Try `torch.Tensor: laf_out with shape [BxNx2x3]`